### PR TITLE
rewrite math in code to match paper

### LINKED ISSUE #368: failure case for exponential cone
```julia
using Pajarito, MathProgBase, CPLEX, ECOS

# min -t
# s.t. (t,s,r)     ∈ Exp
#      1/2 - t - s ≥ 0
# s Binary

c = [-1.0,0.0,0.0]
A = [1.0 1.0 0.0]
b = [0.5]
constr_cones = [(:NonNeg,1:1)]
var_cones = [(:ExpPrimal,1:3)]

function solve(solver)

    m = MathProgBase.ConicModel(solver)
    MathProgBase.loadproblem!(m, c, A, b, constr_cones, var_cones)
    MathProgBase.setvartype!(m, [:Cont,:Bin,:Cont])
    MathProgBase.optimize!(m)
    stat = MathProgBase.status(m)
    x_sol = MathProgBase.getsolution(m)
    objval = MathProgBase.getobjval(m)

    @show stat
    @show x_sol
    @show objval
end


solve(PajaritoSolver(mip_solver=CplexSolver(CPX_PARAM_SCRIND=0),prim_cuts_only=true,solve_relax=false,solve_subp=false))
solve(PajaritoSolver(mip_solver=CplexSolver(CPX_PARAM_SCRIND=0),cont_solver=ECOSSolver(verbose=0),prim_cuts_assist=false))
```

Primal only happy returns an infeasible solution
```
Iter. | Best feasible  | Best bound     | Rel. gap    | Time (s)   
    1 |  -5.000000e-01 |  -5.000000e-01 |   0.000e+00 |   4.680e+00

Iterative algorithm summary:
 - Status               =        Optimal
 - Best feasible        =  -5.000000e-01
 - Best bound           =  -5.000000e-01
 - Relative opt. gap    =      0.000e+00
 - Total time (s)       =       4.72e+00

stat = :Optimal
x_sol = [0.5,0.0,4.0]
objval = -0.5
```

Subp only returns an infeasibile solution but reports issues:
```
Iter. | Best feasible  | Best bound     | Rel. gap    | Time (s)   
    1 |  -5.000000e-01 |  -4.827854e-01 |  -3.443e-02 |   2.096e+00
WARNING: Solution value (-0.5) is smaller than best bound (-0.4827854111813981): check solution feasibility (tightening primal feasibility tolerance of conic solver may help)

Iterative algorithm summary:
 - Status               =        Optimal
 - Best feasible        =  -5.000000e-01
 - Best bound           =  -4.827854e-01
 - Relative opt. gap    =*    -3.443e-02*
 - Total time (s)       =       2.10e+00

Timers (s):
 - Setup                =   3.19e-01
 -- Transform data      =   4.42e-05
 -- Create conic data   =   3.19e-01
 -- Create MIP data     =   6.41e-05
 - Algorithm            =   1.78e+00
 -- Solve relaxation    =   1.09e+00
 -- Get relaxation cuts =   6.02e-01
 -- Solve MIP models    =   8.09e-04
 -- Solve subproblems   =   3.18e-04
 -- Get subproblem cuts =   5.04e-02
 -- Get primal cuts     =   7.83e-06

Counters:
 - Iterations           =     1
 -- Integer repeats     =     0
 -- Conic subproblems   =     1
 --- Infeasible         =     0
 --- Optimal            =     1
 --- Suboptimal         =     0
 --- UserLimit          =     0
 --- ConicFailure       =     0
 --- Other status       =     0
 -- Feasible solutions  =     2
 --- From subproblems   =     1
 --- From MIP solve     =     1

Solution returned by MIP solver

Outer-approximation cuts added:
Cone             | Relax.    | Violated  | Nonviol. 
   Primal expon. |         0 |         1 |         0

0 numerically unstable cone duals encountered

Distance to feasibility (negative indicates strict feasibility):
Cone             | Variable  | Constraint
          Linear |        NA | -0.00e+00

Distance to integrality of integer/binary variables:
          binary |  0.00e+00

stat = :Optimal
x_sol = [0.5,0.0,4.0]
objval = -0.5
```

First thing is to fix the feasibility detection, second is to add the new separation cuts.

### PR DESCRIPTION
**Pull request by chriscoey**
fixes partially #368, but the failure case tests are commented out rn because they do not pass with GLPK 

brings the math in the code in line with the math in the MPC paper

### CODE REVIEW

File: src/conic_algorithm.jl [hunk 1]
```diff
@@ -106,12 +108,12 @@ type PajaritoConicModel <: MathProgBase.AbstractConicModel
...
-    v_idxs_soc_subp::Vector{Vector{Int}} # Row indices of v variables in SOCs in subproblem
-    t_soc::Vector{JuMP.AffExpr} # t variable (epigraph) in SOCs
-    v_soc::Vector{Vector{JuMP.AffExpr}} # v variables in SOCs
-    d_soc::Vector{Vector{JuMP.Variable}} # d variables (disaggregated) in SOCs
-    a_soc::Vector{Vector{JuMP.Variable}} # a variables (absolute values) in SOCs
+    r_idx_soc_subp::Vector{Int} # Row index of r variable in SOCs in subproblems
+    t_idx_soc_subp::Vector{Vector{Int}} # Row indices of t variables in SOCs in subproblem
+    r_soc::Vector{JuMP.AffExpr} # r variable (epigraph) in SOCs
+    t_soc::Vector{Vector{JuMP.AffExpr}} # t variables in SOCs
+    pi_soc::Vector{Vector{JuMP.Variable}} # pi variables (disaggregated) in SOCs
```
Comment (mlubin) [inline src/conic_algorithm.jl]: Do we want to use π?

**Comment by mlubin** [comment]
Agreed that this will make comparing the code and the paper much easier

File: src/conic_algorithm.jl (hunk 1 above)
Comment (chriscoey) [inline src/conic_algorithm.jl]: no not for now, sometimes it doesn't display nicely in some editors (eg not fixed width)

File: src/conic_algorithm.jl [hunk 2]
```diff
@@ -2098,44 +2095,53 @@ function add_cut_sdp!(m, T, W_eig)
...
-            y = T[i,i]
-            z = sum(T[k,l]*mat_dual[k,l] for k in 1:dim, l in 1:dim if (k!=i && l!=i))
-            @constraint(m.model_mip, z >= 0)
-            @expression(m.model_mip, x[j in 1:num_eig], sum((T[k,i]*W_eig[k,j]) for k in 1:dim if k!=i))
-            @expression(m.model_mip, cut_expr, y + z - norm([(y - z), (2*x)...]))
+            # Using SDP SOC full cut (cannot add SOC cuts during callbacks)
+            # Over all diagonal entries i, exclude the largest one
+            i = findmax(abs.(diag(W)))[2]
+
+            # 3-dim rotated-SOC K* constraint is (T_i,i, <T_-i,-i, W_-i,-i>, sqrt2*<T_-i,i, W_eig_1_-i>, ..., sqrt2*<T_-i,i, W_eig_L_-i>) in RSOC^{L+2}, where L is the number of eigenvectors (num_eig)
```
Comment (chriscoey) [inline src/conic_algorithm.jl]: @mlubin this doesn't match the paper

File: src/conic_algorithm.jl [hunk 3]
```diff
@@ -2098,44 +2095,53 @@ function add_cut_sdp!(m, T, W_eig)
...
+                # Using SDP SOC eig cuts (cannot add SOC cuts during callbacks)
                 # Over all diagonal entries i, exclude the largest one
                 i = findmax(abs.(W_eig_j))[2]
 
-                y = T[i,i]
-                z = sum(T[k,l]*W_eig_j[k]*W_eig_j[l] for k in 1:dim, l in 1:dim if (k!=i && l!=i))
-                @constraint(m.model_mip, z >= 0)
-                x = sum(T[k,i]*W_eig_j[k] for k in 1:dim if k!=i)
-                @expression(m.model_mip, cut_expr, y + z - norm([(y - z), 2*x]))
+                # 3-dim rotated-SOC K* constraint is (T_i,i, <T_-i,-i, (W_eig_j*W_eig_j')_-i,-i>, sqrt2*<T_-i,i, W_eig_j_-i>) in RSOC^3
```
Comment (chriscoey) [inline src/conic_algorithm.jl]: @mlubin this doesn't match the paper (eig soc cut case)

File: src/conic_algorithm.jl (hunk 2 above)
Comment (mlubin) [inline src/conic_algorithm.jl]: Comment should be "(L+2)-dim rotated-SOC ..."

**Comment by chriscoey** [comment]
these tests are a little fragile on GLPK unfortunately. CPLEX is fine

**Pull request merged**